### ISSUE TITLE
[FR] Batch process subassembly builds

### ISSUE THREAD

**Issue by camiller000** [inventree/InvenTree#4212]
### Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### Problem statement

We have a part with multiple subassemblies which in turn have multiple levels of subassemblies.  At this time we do not build any of those subassemblies separate and apart from the main build.  We build all parts as we work through the whole build.

In order to allocate stock for all parts within the build, we have to do a build for all of the child builds. Doing a build for each child level is time consuming and requires manually keeping track of the child build creation process to make sure we do build for each subassembly at each level.  

### Suggested solution

Add a feature that would allow a build to be created for all subassemblies that are part of the BOM for the top level part.  This would be similar to the way you can specific the BOM report to include all levels ( (e.g. 0 = all levels).

### Describe alternatives you've considered

We could flatten the BOM to have no subassemblies but that would be unruly

### Do you want to develop this?

- [X] I want to develop this.

**Comment by matmair** [comment]
I think this option would make sense in core. Especially with the possibility to first prototype a few times and then start mass manufacturing.
The changes for this will probably be a bit involved so it might take a bit.

**Comment by SchrodingersGat** [comment]
@camiller000 a good idea, I think this would save a lot of time. Will require a lot of code both front-end and back-end but happy to support it :)

**Comment by SchrodingersGat** [comment]
@temmeand I see the point and I run into this with my technical team at work also. The "can build" is a flat calculation of the top level only. Digging any further down would be computationally expensive (prohibitively so). But perhaps there is a way to approach this?

**Comment by Rincewind34** [comment]
I'm running into a similar issue. The sub builds are really not ideal right now. Even though the is a button to create the sub build in the correct amount (which might be time consuming but is at least convenient and thought out, especially when there are not many), the pain starts when you have sub-sub-builds - especially ones that require the same part to be built over different sub-builds.
It would be truly awesome to have that mechanic built in, that just shows me "You are missing (transitively) 100 self soldered cables for this build". At the moment, this is only possible after submitting all sub-builds and then checking the part screen for the number of missing BOM allocations (and if the part is also used in sub-sub-builds, there is no way to use InvenTree to calculate that - as far as I know).

_Just wanna say again, great software!! Everyone in my team is very pleased with it - keep it up!_

**Comment by SchrodingersGat** [comment]
@Rincewind34 for a high level BOM (with potentially multiple sub-levels of builds) this could get expensive (in terms of database hits) to calculate.

What about the following suggestion:

When creating a new Build, there's an  option for "automatically generate sub builds", which, if selected, would do the following:

Take the parent BOM, and iterate down through the BOM items, looking for how many are "missing" for the BOM. If they are an "assembly", create a new build order (as a child of the parent build order) and then again drill down into that assembly and so on and so on.

This would all be handled by the background worker process, so the user experience would not be slowed.

Thoughts?

**Comment by SchrodingersGat** [comment]
100% agreed that presenting the user with the actions, and allowing them to tweak it, is a better idea. Perhaps this is all achievable in the "front end" with some more consideration

**Comment by camiller000** [comment]
Sorry for long pause.  This change as described above would be meet our needs!

**Comment by walkermc20** [comment]
Really interested in this as well.

**Comment by cbenhagen** [comment]
@SchrodingersGat you recently added the [Auto Create Child Builds Plugin](https://github.com/inventree/InvenTree/pull/9574) which creates builds for sub assemblies with`BuildStatus.PENDING` so if I understand correctly it would not yet trigger new `BuildEvents.ISSUED` events and therefore not work recursively yet? Would it make sense to change that?

**Comment by SchrodingersGat** [comment]
@cbenhagen yes this was an intentional change - automatically generating a top-to-bottom set of build orders proved to be a bit of an "anti-pattern" - in a lot of cases the automatically generated quantities were incorrect (based on other requirements) and needed to be manually edited. 

Additionally changing it to trigger from the "ISSUED" event ensured that the system did not generate dozens of child build orders for a pending order which may end up being cancelled.

In short, the plugin has been adjusted to reduced the amount of "decision making" that InvenTree forces upon you. If you would like a more "fully automated" approach, I would suggest a custom plugin to do this :)

**Comment by cbenhagen** [comment]
@SchrodingersGat I am happy to create our own custom plugin if needed. If possible I would prefer to find a solution that works for others as well though. If recursively walking sub assemblies is slow or error prone let's analyse and fix that.

Were the generated quantities just incorrect because they didn't take existing stock or build batch sizes into account?

**Comment by SchrodingersGat** [comment]
> Were the generated quantities just incorrect because they didn't take existing stock or build batch sizes into account?

It was more a question of how the quantities interacted with other production planning. Individually the quantities were "correct"

**Comment by cbenhagen** [comment]
Can you maybe give me some examples so I can take them into account when thinking about a solution?

**Comment by SchrodingersGat** [comment]
@camiller000  @cbenhagen this has now been implemented as a built-in plugin: https://docs.inventree.org/en/stable/plugins/builtin/auto_create_builds/

**Comment by davidUKYFU** [comment]
@SchrodingersGat 
Hello, I know this issue is old, but what you suggested is exactly what I need. The problem with the current implementation of the Auto Build plugin lies in two points:

1. When you create a build order for an assembly with many subassembly levels, the plugin generates suborders only for the next level. This means you have to enter each order, issue it, and only then does it generate the next layers.

2. It would also be helpful to have a tree view of the subassembly parts in the "Required Parts" section, so the user doesn’t need to open each build order to assign stock and check requirements.

**Comment by SchrodingersGat** [comment]
@davidUKYFU the functionality was intentionally reduced here, as the default behaviour automatically spawning many multi-level build orders was too opaque - at least for default behaviour.

I would suggest that this could be implemented as a custom plugin, if you wanted?

**Comment by davidUKYFU** [comment]
@SchrodingersGat Yes, making the build orders for all the subassamblies could be just a mod of the actual implementation. However, the part about viewing the assembly’s BOM in a tree structure from the Required Materials screen might require something beyond a simple plugin.
